File: java/146.lru-cache.java
/*
 * @lc app=leetcode id=146 lang=java
 *
 * [146] LRU Cache
 */

// @lc code=start
class LRUCache {

    class Node {
        Node prev;
        Node next;
        int val;

        public Node() {
            this.prev = null;
            this.next = null;
            this.val = 0;
        }
    }

    int capacity;
    int size;
    Node sentinel;

    Map<Integer, Node> nodeMap;
    Map<Integer, Integer> kvMap;


    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.size = 0;

        this.sentinel = new Node();
        this.sentinel.next = this.sentinel;
        this.sentinel.prev = this.sentinel;

        this.nodeMap = new HashMap();
        this.kvMap = new HashMap();
    }
    
    public int get(int key) {
        int ret = kvMap.getOrDefault(key, -1);

        if(ret != -1) {
            Node n = nodeMap.get(key);
            remove(n);
            add(n);
        }

        return ret;
    }
    
    public void put(int key, int value) {
        Node n = nodeMap.get(key);
        if(n == null) {
            if(size == capacity) {
                nodeMap.remove(this.sentinel.prev.val);
                kvMap.remove(this.sentinel.prev.val);
                remove(this.sentinel.prev);
            }
            else {
                size++;
            } 

            n = new Node();
            n.val = key;
        }
        else {
            remove(n);
        }

        kvMap.put(key, value);
        nodeMap.put(key, n);
        add(n);
    }

    private void remove(Node n) {
        n.prev.next = n.next;
        n.next.prev = n.prev;
    }

    private void add(Node n) {
        n.next = this.sentinel.next;
        this.sentinel.next.prev = n;

        n.prev = this.sentinel;
        this.sentinel.next = n;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
// @lc code=end

File: cpp/146.lru-cache.cpp
/*
 * @lc app=leetcode id=146 lang=cpp
 *
 * [146] LRU Cache
 */

// @lc code=start
class LRUCache {
public:
    LRUCache(int capacity) : capacity_(capacity) {}

    int get(int key) {
        auto it = map_.find(key);
        if (it == map_.end()) {
            return -1; 
        }
        list_.splice(list_.begin(), list_, it->second);
        
        return it->second->second;
    }

    void put(int key, int value) {
        auto it = map_.find(key);
        
        if (it != map_.end()) {
            it->second->second = value;
            list_.splice(list_.begin(), list_, it->second);

            return;
        }

        if (list_.size() == capacity_) {
            map_.erase(list_.back().first);
            list_.pop_back();
        }

        list_.emplace_front(key, value);
        map_[key] = list_.begin();
    }

private:
    int capacity_;
    std::list<std::pair<int, int>> list_;
    std::unordered_map<int, std::list<std::pair<int, int>>::iterator> map_;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end

